### PR TITLE
Permute the dofs on quads and hexes

### PR DESCRIPTION
**Pull request by mscroggs**
This pull request adds the necessary permutations to `dof_permutation.cpp` for quads and hexes.

Currently DG (https://github.com/FEniCS/dolfinx/issues/645) and vector function spaces (https://github.com/FEniCS/dolfinx/issues/31) do not work unless the mesh is ordered, `Ordering::order_simplex` should be called in these cases.

### CODE REVIEW

File: cpp/dolfin/fem/dofs_permutation.cpp [hunk 1]
```diff
@@ -221,22 +423,87 @@ compute_ordering_triangle(const mesh::Mesh& mesh)
...
       cell_orders(num_cells, num_permutations);
 
+  const std::vector<std::int64_t>& global_indices
+      = mesh.topology().global_indices(0);
   // Set orders for each cell
   for (int cell_n = 0; cell_n < num_cells; ++cell_n)
   {
     const mesh::MeshEntity cell(mesh, 2, cell_n);
     const std::int32_t* vertices = cell.entities(0);
+    const std::int32_t v0 = global_indices[vertices[0]];
```
Comment (garth-wells) [inline cpp/dolfin/fem/dofs_permutation.cpp]: Global indices needs to be 64-bit ints.

File: cpp/dolfin/fem/dofs_permutation.cpp [hunk 2]
```diff
@@ -221,22 +423,87 @@ compute_ordering_triangle(const mesh::Mesh& mesh)
...
+      cell_orders(num_cells, num_permutations);
+
+  // Set orders for each cell
+  const std::vector<std::int64_t>& global_indices
+      = mesh.topology().global_indices(0);
+  for (int cell_n = 0; cell_n < num_cells; ++cell_n)
+  {
+    const mesh::MeshEntity cell(mesh, 1, cell_n);
+    const std::int32_t* vertices = cell.entities(0);
+    const std::int32_t v0 = global_indices[vertices[0]];
```
Comment (garth-wells) [inline cpp/dolfin/fem/dofs_permutation.cpp]: Needs to be `std::int64_t`.

File: cpp/dolfin/fem/dofs_permutation.cpp [hunk 3]
```diff
@@ -221,22 +423,87 @@ compute_ordering_triangle(const mesh::Mesh& mesh)
...
+      cell_orders(num_cells, num_permutations);
+
+  // Set orders for each cell
+  const std::vector<std::int64_t>& global_indices
+      = mesh.topology().global_indices(0);
+  for (int cell_n = 0; cell_n < num_cells; ++cell_n)
+  {
+    const mesh::MeshEntity cell(mesh, 2, cell_n);
+    const std::int32_t* vertices = cell.entities(0);
+    const std::int32_t v0 = global_indices[vertices[0]];
```
Comment (garth-wells) [inline cpp/dolfin/fem/dofs_permutation.cpp]: Needs to be `std::int64_t`.

**Comment by mscroggs** [comment]
> Are global indices needed to match up the dofs on processed boundaries?
Global indices are needed so that the same edge viewed from two parallel processes will be give the same ordering

File: cpp/dolfin/fem/dofs_permutation.cpp [hunk 4]
```diff
@@ -83,6 +91,89 @@ std::array<int, 4> calculate_tetrahedron_orders(int v1, int v2, int v3, int v4)
...
 //-----------------------------------------------------------------------------
+/// Calculates the number of times the rotation and reflection of a
+/// quadrilateral should be applied to a quadrilateral with the given global
+/// vertex numbers
+/// @param[in] v1 The global vertex number of the quadrilateral's first vertex
+/// @param[in] v2 The global vertex number of the quadrilateral's second vertex
+/// @param[in] v3 The global vertex number of the quadrilateral's third vertex
+/// @param[in] v4 The global vertex number of the quadrilateral's fourth vertex
+/// @return The rotation and reflection orders for the quadrilateral
+std::array<int, 2> calculate_quadrilateral_orders(int v1, int v2, int v3,
```
Comment (garth-wells) [inline cpp/dolfin/fem/dofs_permutation.cpp]: This needs updating to 64-bit int (and others below).

File: cpp/dolfin/fem/dofs_permutation.cpp [hunk 5]
```diff
@@ -83,6 +91,89 @@ std::array<int, 4> calculate_tetrahedron_orders(int v1, int v2, int v3, int v4)
...
+/// Calculates the number of times the rotations and reflection of a triangle
+/// should be applied to a hexahedron with the given global vertex numbers
+/// @param[in] v1 The global vertex number of the hexahedron's first vertex
+/// @param[in] v2 The global vertex number of the hexahedron's second vertex
+/// @param[in] v3 The global vertex number of the hexahedron's third vertex
+/// @param[in] v4 The global vertex number of the hexahedron's fourth vertex
+/// @param[in] v5 The global vertex number of the hexahedron's fifth vertex
+/// @param[in] v6 The global vertex number of the hexahedron's sixth vertex
+/// @param[in] v7 The global vertex number of the hexahedron's seventh vertex
+/// @param[in] v8 The global vertex number of the hexahedron's eighth vertex
```
Comment (michalhabera) [inline cpp/dolfin/fem/dofs_permutation.cpp]: Wow this is verbose...

Comment (mscroggs) [reply cpp/dolfin/fem/dofs_permutation.cpp]: Yes, I'll change it to "v1 to v8 are..."

File: cpp/dolfin/fem/dofs_permutation.cpp (hunk 4 above)
Comment (mscroggs) [reply cpp/dolfin/fem/dofs_permutation.cpp]: I've added templating to (I think) everywhere that a larger or smaller int is needed

**Pull request merged**